h2o-core/src/main/java/water/jdbc/SQLManager.java

```diff
@@ -171,39 +185,62 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
+
     if (optimize) {
-      _v = makeCon(totSize, numRow);
+      // for optimal retrieval and rebalancing, use min 1 chunk per node, and max chunks = min(max(total threads), max(total connections))
+      final int num_retrieval_chunks = H2O.getCloudSize() * ConnectionPoolProvider.getOptimalConnectionPerNode(H2O.getCloudSize(), H2O.ARGS.nthreads);
+      _retrieval_v = num_retrieval_chunks >= _v.nChunks() ? _v : Vec.makeConN(numRow, num_retrieval_chunks);
     } else {
-      double rows_per_chunk = FileVec.calcOptimalChunkSize(totSize, numCol, numCol * 4,
-              Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false, false);
-      _v = makeCon(0, numRow, (int) Math.ceil(Math.log1p(rows_per_chunk)), false);
+      _retrieval_v = _v;
     }
-    Log.info("Number of chunks: " + _v.nChunks());
+    //if autoRebalance set to true, we first use optimal #chunks for retrieval and then immediately rebalance to optimal #chunks for later processing
+    final boolean autoRebalance = _v != _retrieval_v && Boolean.parseBoolean(System.getProperty(AUTO_REBALANCE_ENABLED, "false"));
+    Log.info("Number of chunks for data retrieval: " + _retrieval_v.nChunks());
+    Log.info("Number of final chunks: " + (autoRebalance ? _v.nChunks() : _retrieval_v.nChunks()));
     //create frame
-    final Key destination_key = Key.make(table + "_sql_to_hex");
+    final Key destination_key = Key.make((table + "_sql_to_hex").replaceAll("\\W", "_"));
     final Job<Frame> j = new Job(destination_key, Frame.class.getName(), "Import SQL Table");
 
     final String finalTable = table;
     H2O.H2OCountedCompleter work = new H2O.H2OCountedCompleter() {
       @Override
       public void compute2() {
-        ConnectionPoolProvider provider = new ConnectionPoolProvider(connection_url, username, password, _v.nChunks());
-        Frame fr = new SqlTableToH2OFrame(finalTable, needFetchClause, columns, numCol, j, provider).doAll(columnH2OTypes, _v)
-                .outputFrame(destination_key, columnNames, null);
-        DKV.put(fr);
+        final ConnectionPoolProvider provider = new ConnectionPoolProvider(connection_url, username, password, _retrieval_v.nChunks());
+        final Frame fr;
+        final Key tmpKey = Key.make();
+        final Frame retrieval_fr = new SqlTableToH2OFrame(finalTable, needFetchClause, columns, numCol, j, provider)
+                .doAll(columnH2OTypes, _retrieval_v)
+                .outputFrame(tmpKey, columnNames, null);
+
+        if (autoRebalance) {
+          final RebalanceDataSet rds = new RebalanceDataSet(retrieval_fr, destination_key, _v.nChunks());
+          H2O.submitTask(rds);
+          fr = rds.getResult();
+        } else {
+          fr = new Frame(destination_key, retrieval_fr.names(), retrieval_fr.vecs());
+        }
+        Frame.deleteTempFrameAndItsNonSharedVecs(retrieval_fr, fr);
+        _retrieval_v.remove();
         _v.remove();
```

Comment:
Is Vec `_v` used for anything? I am probably missing something but it looks like it is merely holding number of chunks and it is not used as a Vec.